### PR TITLE
optional param

### PR DESCRIPTION
**Pull request by hochang**

### CODE REVIEW

**Comment by shimonchayim** [comment]
enhancement required for spec here: https://github.com/ql-io/ql.io/wiki/%5BProposal%5D-Optional-Inputs-and-Errors

**Comment by hochang** [comment]
Changes added to pull request :)
Thank you and best regards
Hongcheng Chang

On 5/10/12 12:08 PM, "shimonchayim"
<reply+i-4517554-dd8aa24186d4226a313f7d7466f11d6673173d43-1302440@reply.git
hub.com> wrote:

**Comment by shimonchayim** [comment]
Good work ..We can review the code on Monday and will merge the pull request.

File: modules/compiler/pegs/ql.peg [hunk 1]
```diff
@@ -990,15 +990,20 @@ RetRef = o:Word {
     }
 }
 
-Route = 'via' insig 'route' insig p:QuotedWord insig 'using' insig 'method' insig
```
Comment (shimonchayim) [inline modules/compiler/pegs/ql.peg]: We are missing handling of: "using defaults 'p2' = 'some default'" as specified at this link: https://github.com/ql-io/ql.io/wiki/%5BProposal%5D-Optional-Inputs-and-Errors

File: modules/engine/pegs/str-template.peg [hunk 2]
```diff
@@ -137,7 +137,7 @@ template = c:( expression / literal )* {
     }
 }
 
-literal = (([^^\`{}]) / ' ') / expression
```
Comment (hochang) [inline modules/engine/pegs/str-template.peg:140]: Subbu, where else does this literal getting used? Can removing ^ break some other places? Nothing breaks in make test at least.

Comment (s3u) [inline modules/engine/pegs/str-template.peg:140]: Good catch. This is not needed.

File: modules/compiler/pegs/ql.peg (hunk 1 above)
Comment (s3u) [inline modules/compiler/pegs/ql.peg]: Yeah - needs `using defaults` in the grammar. Substituting those details is as trivial as setting defaults on `routeParams`.

File: modules/console/app.js [hunk 3]
```diff
@@ -252,15 +252,39 @@ var Console = module.exports = function(opts, cb) {
...
+                                return _.has(verbRouteVariant.routeInfo.defaults, verbRouteVariant.query[k]);
+                            })
+                            .value();
+                        //missed query params that are neither defaults nor user provided
+                        var missed = _.difference(_.keys(verbRouteVariant.query), _.union(defaultKeys, _.keys(holder.params)));
+                        var misrequired = _.filter(missed, function(key){
+                            if (verbRouteVariant.routeInfo.optparam){
+                                // if with optional params, find if any required param is missed
+                                return verbRouteVariant.query[key] && verbRouteVariant.query[key].indexOf("^") == 0;
+                            }else{
```
Comment (s3u) [inline modules/console/app.js]: Do you mind putting `else {` on the next line to be consistent with the rest of the file?

File: modules/console/app.js [hunk 4]
```diff
@@ -272,6 +296,13 @@ var Console = module.exports = function(opts, cb) {
                     return;
                 }
 
+
+                //collect default query params if needed
```
Comment (s3u) [inline modules/console/app.js]: Also, a space after `//` would be better

File: modules/console/app.js [hunk 5]
```diff
@@ -272,6 +296,13 @@ var Console = module.exports = function(opts, cb) {
...
+                _.defaults(holder.params, route.routeInfo.defaults);
+                //var keys = _.difference(_.keys(route.routeInfo.defaults), _.keys(holder.params))
+                //_.each(keys, function(key){
+                //    holder.params[key] = route.routeInfo.defaults[key];
+                //})
```
Comment (s3u) [inline modules/console/app.js]: Any reason to keep commented out code?

File: modules/console/test/routes/optional-findItems.ql [hunk 6]
```diff
@@ -0,0 +1,3 @@
+-- this route does not have the word 'with optional params', so requires all params to be provided.
+return select itemId from finditems where keywords = '{^keyword}'
```
Comment (s3u) [inline modules/console/test/routes/optional-findItems.ql:2]: <del>word</del>  clause

Comment (hochang) [inline modules/console/test/routes/optional-findItems.ql:2]: Revised, please check again.
Thank you and best regards
Hongcheng Chang

On 5/16/12 10:52 AM, "Subbu Allamaraju"
<reply+p-833391-d3749f2cf8ea7cc954fc3a49ef7f1f123c9046dd-1302440@reply.gith
ub.com> wrote:

**Comment by s3u** [comment]
Please also bump up version numbers in package.json of effected modules and the README.md

**Pull request merged**